### PR TITLE
[server] Add OTel metrics to RocksDBStats

### PR DESCRIPTION
**Pull request by m-nagarajan**
## Problem Statement

RocksDBStats (gated by `rocksdb.statistics.enabled`, default false) reports 27 Tehuti AsyncGauge sensors for RocksDB ticker counters but has no OTel metrics. This prevents migration to OTel-based monitoring for RocksDB block cache, memtable, bloom filter, compaction, and read amplification statistics.

## Solution

Add 13 OTel metrics alongside the existing 27 Tehuti sensors, consolidating per-component metrics using dimensions:

- **4 per-component block cache metrics** (`miss.count`, `hit.count`, `add.count`, `bytes.inserted`) with `VENICE_ROCKSDB_BLOCK_CACHE_COMPONENT` dimension (INDEX/FILTER/DATA/COMPRESSION_DICT) — consolidates 16 Tehuti sensors into 4 OTel metrics
- **7 joint Tehuti+OTel metrics** for block cache add failures, bytes read/write, bloom filter, memtable hit/miss, and compaction cancelled
- **1 per-level get-hit metric** (`get.hit.count`) with `VENICE_ROCKSDB_LEVEL` dimension (LEVEL_0/LEVEL_1/LEVEL_2_AND_UP) — consolidates 3 Tehuti sensors into 1 OTel metric
- **1 read amplification factor** using `ASYNC_DOUBLE_GAUGE` with split callbacks (Tehuti returns -1 sentinel, OTel returns NaN when unavailable so SDK drops data point)

Overall block cache miss/hit/add counters remain Tehuti-only — OTel totals are derivable via `sum by (component)`. Block cache hit ratio is also Tehuti-only (derivable from per-component OTel gauges).

<details>
<summary>Tehuti → OTel metric mapping (click to expand)</summary>

| Tehuti Sensor | OTel Metric | OTel Dimensions | Notes |
|---|---|---|---|
| `rocksdb_block_cache_index_miss` | `rocksdb.block_cache.miss.count` | COMPONENT=INDEX | Per-component |
| `rocksdb_block_cache_filter_miss` | `rocksdb.block_cache.miss.count` | COMPONENT=FILTER | Per-component |
| `rocksdb_block_cache_data_miss` | `rocksdb.block_cache.miss.count` | COMPONENT=DATA | Per-component |
| `rocksdb_block_cache_compression_dict_miss` | `rocksdb.block_cache.miss.count` | COMPONENT=COMPRESSION_DICT | Per-component |
| `rocksdb_block_cache_index_hit` | `rocksdb.block_cache.hit.count` | COMPONENT=INDEX | Per-component |
| `rocksdb_block_cache_filter_hit` | `rocksdb.block_cache.hit.count` | COMPONENT=FILTER | Per-component |
| `rocksdb_block_cache_data_hit` | `rocksdb.block_cache.hit.count` | COMPONENT=DATA | Per-component |
| `rocksdb_block_cache_compression_dict_hit` | `rocksdb.block_cache.hit.count` | COMPONENT=COMPRESSION_DICT | Per-component |
| `rocksdb_block_cache_index_add` | `rocksdb.block_cache.add.count` | COMPONENT=INDEX | Per-component |
| `rocksdb_block_cache_filter_add` | `rocksdb.block_cache.add.count` | COMPONENT=FILTER | Per-component |
| `rocksdb_block_cache_data_add` | `rocksdb.block_cache.add.count` | COMPONENT=DATA | Per-component |
| `rocksdb_block_cache_compression_dict_add` | `rocksdb.block_cache.add.count` | COMPONENT=COMPRESSION_DICT | Per-component |
| `rocksdb_block_cache_index_bytes_insert` | `rocksdb.block_cache.bytes.inserted` | COMPONENT=INDEX | Per-component |
| `rocksdb_block_cache_filter_bytes_insert` | `rocksdb.block_cache.bytes.inserted` | COMPONENT=FILTER | Per-component |
| `rocksdb_block_cache_data_bytes_insert` | `rocksdb.block_cache.bytes.inserted` | COMPONENT=DATA | Per-component |
| `rocksdb_block_cache_compression_dict_bytes_insert` | `rocksdb.block_cache.bytes.inserted` | COMPONENT=COMPRESSION_DICT | Per-component |
| `rocksdb_block_cache_add_failures` | `rocksdb.block_cache.add.failure.count` | — | Joint |
| `rocksdb_block_cache_bytes_read` | `rocksdb.block_cache.bytes.read` | — | Joint |
| `rocksdb_block_cache_bytes_write` | `rocksdb.block_cache.bytes.written` | — | Joint |
| `rocksdb_bloom_filter_useful` | `rocksdb.bloom_filter.useful_count` | — | Joint |
| `rocksdb_memtable_hit` | `rocksdb.memtable.hit.count` | — | Joint |
| `rocksdb_memtable_miss` | `rocksdb.memtable.miss.count` | — | Joint |
| `rocksdb_compaction_cancelled` | `rocksdb.compaction.cancelled_count` | — | Joint |
| `rocksdb_get_hit_l0` | `rocksdb.get.hit.count` | LEVEL=LEVEL_0 | Per-level |
| `rocksdb_get_hit_l1` | `rocksdb.get.hit.count` | LEVEL=LEVEL_1 | Per-level |
| `rocksdb_get_hit_l2_and_up` | `rocksdb.get.hit.count` | LEVEL=LEVEL_2_AND_UP | Per-level |
| `rocksdb_read_amplification_factor` | `rocksdb.read_amplification_factor` | — | ASYNC_DOUBLE_GAUGE |
| `rocksdb_block_cache_miss` | — | — | Tehuti-only (OTel derives from sum) |
| `rocksdb_block_cache_hit` | — | — | Tehuti-only (OTel derives from sum) |
| `rocksdb_block_cache_add` | — | — | Tehuti-only (OTel derives from sum) |
| `rocksdb_block_cache_hit_ratio` | — | — | Tehuti-only (derivable from OTel gauges) |

</details>

Dimension-to-TickerType mappings use static `EnumMap`s. A missing mapping causes the `liveStateResolver` to return `null` (dormant — no data point emitted) rather than throwing; completeness is enforced by per-enum CI tests (`testEveryVeniceRocksDBLevelEmitsForGetHitCount`, `testEveryVeniceRocksDBBlockCacheComponentEmitsForAllPerComponentMetrics`) so adding a new `VeniceRocksDBLevel` or `VeniceRocksDBBlockCacheComponent` value without updating the ticker map fails CI rather than spamming failure metrics every collection cycle.

Also fixes: `AggRocksDBStats` lambda was capturing the outer `metricsRepository` constructor parameter instead of the lambda parameter `metricsRepo`, and adds null guard to `setRocksDBStat`. `rocksDBStat` field made `volatile` for JMM correctness. `RocksDBStats` does NOT propagate `isTotalStats()` to `OpenTelemetryMetricsSetup` — `AggRocksDBStats` is fire-and-forget in `VeniceServer.java`, so the `totalStats` instance is the only `RocksDBStats` ever created in production. Propagating `isTotalStats(true)` would suppress every RocksDB OTel emission. This matches the pattern used by `RocksDBMemoryStats` (the closest single-instance analog).

### Code changes
- [ ] Added new code behind **a config**. If so list the config names and their default values in the PR description.
- [ ] Introduced new **log lines**.
  - [ ] Confirmed if logs need to be **rate limited** to avoid excessive logging.

### **Concurrency-Specific Checks**
Both reviewer and PR author to verify
- [x] Code has **no race conditions** or **thread safety issues**.
- [x] Proper **synchronization mechanisms** (e.g., `synchronized`, `RWLock`) are used where needed.
- [x] No **blocking calls** inside critical sections that could lead to deadlocks or performance degradation.
- [x] Verified **thread-safe collections** are used (e.g., `ConcurrentHashMap`, `CopyOnWriteArrayList`).
- [x] Validated proper exception handling in multi-threaded code to avoid silent thread termination.

## How was this PR tested?

- [x] New unit tests added.
- [ ] New integration tests added.
- [ ] Modified or extended existing tests.
- [x] Verified backward compatibility (if applicable).

New test files:
- `RocksDBStatsOtelTest` (15 tests): per-component dimension validation for all 4 block cache metrics × 4 components, joint metrics, SST-level dimensions, read amplification `ASYNC_DOUBLE_GAUGE` with NaN edge case, pre-init behavior (joint `AsyncMetricEntityStateBase` metrics emit `-1` sentinel; per-component / per-level `AsyncMetricEntityStateOneEnum` metrics honor the dormant contract and emit no data point), per-enum completeness guards (`testEveryVeniceRocksDBLevelEmitsForGetHitCount`, `testEveryVeniceRocksDBBlockCacheComponentEmitsForAllPerComponentMetrics`), `isTotalStats` regression guard (`testOtelEmitsForTotalStatsName` constructs `RocksDBStats(repo, "total", ...)` and asserts emission so re-introducing `isTotalStats()` propagation fails CI), NPE prevention (OTel disabled + plain `MetricsRepository`). Tests use a dedicated `AsyncGauge.AsyncGaugeExecutor` per `VeniceMetricsRepository` so `close()` doesn't shut down the JVM-wide singleton executor and break later AsyncGauge tests.
- `RocksDBStatsTehutiTest` (5 tests): Tehuti sensor baseline (ticker values, hit ratio formula, read amplification -1 sentinel, NaN on zero denominator, setRocksDBStat guards)
- `RocksDBStatsOtelMetricEntityTest`: validates all 13 entity definitions via ModuleMetricEntityTestFixture
- `VeniceRocksDBBlockCacheComponentTest` + `VeniceRocksDBLevelTest`: dimension enum validation via VeniceDimensionTestFixture
- `VeniceMetricsDimensionsTest`: updated for both new dimensions across all 3 naming formats
- `ServerMetricEntityTest`: count updated from 159 to 172

## Does this PR introduce any user-facing or breaking changes?
- [x] No. You can skip the rest of this section.
- [ ] Yes. Clearly explain the behavior change and its impact.

### CODE REVIEW

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/RocksDBStatsOtelMetricEntity.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+      "rocksdb.block_cache.add.count", MetricUnit.NUMBER, "Blocks added to cache by component",
+      setOf(VENICE_CLUSTER_NAME, VENICE_ROCKSDB_BLOCK_CACHE_COMPONENT)
+  ),
+
+  BLOCK_CACHE_BYTES_INSERTED(
+      "rocksdb.block_cache.bytes.inserted", MetricUnit.BYTES, "Bytes inserted into cache by component",
+      setOf(VENICE_CLUSTER_NAME, VENICE_ROCKSDB_BLOCK_CACHE_COMPONENT)
+  ),
+
+  // Block Cache — overall (no per-component breakdown available)
```
Comment (m-nagarajan) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/RocksDBStatsOtelMetricEntity.java:55]: can rocksdb.block_cache.read.bytes -> rocksdb.block_cache.bytes.read
 rocksdb.block_cache.write.bytes -> rocksdb.block_cache.bytes.written

similar to rocksdb.block_cache.bytes.inserted? Other similar one if relevant

File: services/venice-server/src/main/java/com/linkedin/venice/stats/AggRocksDBStats.java [hunk 2]
```diff
@@ -12,7 +12,7 @@ public AggRocksDBStats(String cluster, MetricsRepository metricsRepository, Stat
     super(
         cluster,
         metricsRepository,
-        (metricsRepo, storeName, clusterName) -> new RocksDBStats(metricsRepository, storeName),
+        (metricsRepo, storeName, clusterName) -> new RocksDBStats(metricsRepo, storeName, clusterName),
```
Comment (m-nagarajan) [inline services/venice-server/src/main/java/com/linkedin/venice/stats/AggRocksDBStats.java:15]: 💡 **[SUGGESTION]** No test for `AggRocksDBStats` verifying the lambda and clusterName fix end-to-end

The PR fixes two bugs in `AggRocksDBStats`: the lambda was capturing the outer `metricsRepository` (instead of the lambda parameter `metricsRepo`), and `clusterName` was never passed to `RocksDBStats`. These fixes are correct but have no test. The only guard is `RocksDBStatsTehutiTest` and `RocksDBStatsOtelTest`, which test `RocksDBStats` directly with a `MetricsRepository`/`VeniceMetricsRepository`. An `AggRocksDBStats` test instantiating the class, calling `addStore`, and confirming that the per-store stats use `metricsRepo` (not the outer `metricsRepository`) and have the correct cluster name in OTel attributes would catch any regression in this lambda.

**Suggested fix:**
Add a test that creates `AggRocksDBStats`, calls `getStoreStats("some-store")`, and verifies that the returned `RocksDBStats` instance's OTel attributes include `VENICE_CLUSTER_NAME = "test-cluster"` (i.e., that `clusterName` was correctly threaded through).

**Comment by m-nagarajan** [comment]
**services/venice-server/src/main/java/com/linkedin/venice/stats/RocksDBStats.java:829**

💡 **[SUGGESTION]** Double-lambda wrapper in `registerJointTickerMetric` is unnecessary — the `LongSupplier` can be passed directly to AsyncGauge

The helper method creates a `LongSupplier callback`, then wraps it in a second lambda for Tehuti:

```java
LongSupplier callback = () -> rocksDBStat != null ? rocksDBStat.getTickerCount(tickerType) : -1;
registerSensorIfAbsent(new AsyncGauge((ig, ig2) -> callback.getAsLong(), tehutiSensorName));
```

`AsyncGauge` accepts any `MetricValueProvider`, and the `(ig, ig2) -> callback.getAsLong()` wrapper only exists to bridge `LongSupplier` → `MetricValueProvider`. The same approach is already used in `registerTickerSensor` with a direct inline lambda. The `callback` variable is valid because it's reused for both Tehuti and OTel, but the Tehuti wrapper could be:

```java
registerSensorIfAbsent(new AsyncGauge((ig, ig2) -> callback.getAsLong(), tehutiSensorName));
```

…which is what the code already does, so no change needed. Actually the issue is that the callback could be used directly as an `AsyncGauge` if `AsyncGauge` accepted a `LongSupplier`. Since it doesn't, the current two-lambda approach is the minimal implementation. Minor note only — no code change needed.

**Comment by m-nagarajan** [comment]
**services/venice-server/src/main/java/com/linkedin/venice/stats/RocksDBStats.java:660**

💡 **[SUGGESTION]** Overall block cache `miss`/`hit`/`add` OTel derivability should be documented at the caller site

The three overall block cache counters (`BLOCK_CACHE_MISS`, `BLOCK_CACHE_HIT`, `BLOCK_CACHE_ADD`) are registered as Tehuti-only with a comment "OTel derives totals via `sum by (component)`". The design is sound and documented in the class Javadoc, but the inline comments at the registration sites (lines 661–663) only say "Tehuti-only, OTel derives from per-component sum". For operators migrating alert rules from Tehuti to OTel, knowing the exact PromQL/OTel aggregation query would reduce friction. The class-level Javadoc already describes this, but the inline comment at the registration could reference it.

**Suggested fix:**
Add a short comment at the `registerTickerSensor("rocksdb_block_cache_miss", ...)` etc. lines pointing back to the class Javadoc: `// Tehuti-only; OTel total derivable via: sum(rocksdb.block_cache.miss.count by cluster)`

File: services/venice-server/src/main/java/com/linkedin/venice/stats/RocksDBStats.java [hunk 3]
```diff
@@ -28,112 +45,273 @@
...
+    ADD_TICKER_BY_COMPONENT.put(VeniceRocksDBBlockCacheComponent.COMPRESSION_DICT, BLOCK_CACHE_COMPRESSION_DICT_ADD);
+
+    BYTES_INSERT_TICKER_BY_COMPONENT.put(VeniceRocksDBBlockCacheComponent.INDEX, BLOCK_CACHE_INDEX_BYTES_INSERT);
+    BYTES_INSERT_TICKER_BY_COMPONENT.put(VeniceRocksDBBlockCacheComponent.FILTER, BLOCK_CACHE_FILTER_BYTES_INSERT);
+    BYTES_INSERT_TICKER_BY_COMPONENT.put(VeniceRocksDBBlockCacheComponent.DATA, BLOCK_CACHE_DATA_BYTES_INSERT);
+    BYTES_INSERT_TICKER_BY_COMPONENT
+        .put(VeniceRocksDBBlockCacheComponent.COMPRESSION_DICT, BLOCK_CACHE_COMPRESSION_DICT_BYTES_INSERT);
+  }
+
   private Statistics rocksDBStat;
```
[the remark this answers is not in the document]
Comment (m-nagarajan) [reply services/venice-server/src/main/java/com/linkedin/venice/stats/RocksDBStats.java]: Fixed. Made `rocksDBStat` volatile. While the write-once pattern with -1 sentinel was safe in practice, volatile makes it correct under the JMM and aligns with Venice's rule: "Read volatile references once at the start of a method."

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/RocksDBStatsOtelMetricEntity.java (hunk 1 above)
Comment (m-nagarajan) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/RocksDBStatsOtelMetricEntity.java:55]: Fixed. Renamed to `rocksdb.block_cache.bytes.read` and `rocksdb.block_cache.bytes.written` — all three byte metrics now share `bytes` as the common namespace level: `bytes.inserted`, `bytes.read`, `bytes.written`.

File: services/venice-server/src/main/java/com/linkedin/venice/stats/AggRocksDBStats.java (hunk 2 above)
Comment (m-nagarajan) [reply services/venice-server/src/main/java/com/linkedin/venice/stats/AggRocksDBStats.java:15]: Acknowledged. `AggRocksDBStats` is a 4-line wrapper — the lambda fix and clusterName plumbing are covered indirectly by `RocksDBStatsOtelTest` (validates OTel metrics with cluster dimension) and `RocksDBStatsTehutiTest` (validates Tehuti sensors). Adding an integration test through `AggRocksDBStats` would require mocking `org.rocksdb.Statistics` (JNI class) at the aggregate level, adding complexity for marginal value. The lambda fix is visually obvious in the diff.

**Review by sushantmane — DISMISSED** [review]
LGTM

File: services/venice-server/src/main/java/com/linkedin/venice/stats/RocksDBStats.java [hunk 4]
```diff
@@ -28,112 +45,274 @@
...
-    this.getHitL0 = registerSensor("rocksdb_get_hit_l0", GET_HIT_L0);
-    this.getHitL1 = registerSensor("rocksdb_get_hit_l1", GET_HIT_L1);
-    this.getHitL2AndUp = registerSensor("rocksdb_get_hit_l2_and_up", GET_HIT_L2_AND_UP);
-    this.compactionCancelled = registerSensor("rocksdb_compaction_cancelled", COMPACTION_CANCELLED);
-
-    this.blockCacheHitRatio = registerSensor(new AsyncGauge((ignored, ignored2) -> {
+    OpenTelemetryMetricsSetup.OpenTelemetryMetricsSetupInfo otelData =
+        OpenTelemetryMetricsSetup.builder(metricsRepository)
+            .setClusterName(clusterName)
+            .isTotalStats(isTotalStats())
```
[the remark this answers is not in the document]
Comment (m-nagarajan) [reply services/venice-server/src/main/java/com/linkedin/venice/stats/RocksDBStats.java]: Confirmed and fixed. Traced the production wiring: `AggRocksDBStats` is fire-and-forget at `VeniceServer.java:357` (`.ifPresent(stat -> new AggRocksDBStats(...))`), so `getStoreStats(storeName)` is never called and the `totalStats` instance (named `"total"`) is the only `RocksDBStats` ever created in production. With `isTotalStats()` propagation, every RocksDB OTel metric was suppressed.

Removed the propagation in `RocksDBStats.java`:

```java
// AggRocksDBStats only constructs the "total" stats instance and never per-store stats, so
// isTotalStats() must NOT be propagated — it would suppress all OTel emission in production.
OpenTelemetryMetricsSetup.OpenTelemetryMetricsSetupInfo otelData =
    OpenTelemetryMetricsSetup.builder(metricsRepository).setClusterName(clusterName).build();
```

Matches the pattern used by `RocksDBMemoryStats` (closest single-instance analog — also constructed once per process, not via Agg/per-store split).

Added regression test `testOtelEmitsForTotalStatsName` at `RocksDBStatsOtelTest.java:295` that constructs `RocksDBStats(repo, "total", ...)` and asserts emission of `rocksdb.get.hit.count`, so any reintroduction of `.isTotalStats(isTotalStats())` propagation fails CI.

File: services/venice-server/src/test/java/com/linkedin/venice/stats/RocksDBStatsOtelTest.java [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+
+  @BeforeMethod
+  public void setUp() {
+    inMemoryMetricReader = InMemoryMetricReader.create();
+    metricsRepository = new VeniceMetricsRepository(
+        new VeniceMetricsConfig.Builder().setMetricPrefix(TEST_METRIC_PREFIX)
+            .setMetricEntities(SERVER_METRIC_ENTITIES)
+            .setEmitOtelMetrics(true)
+            .setOtelAdditionalMetricsReader(inMemoryMetricReader)
+            .build());
```
[the remark this answers is not in the document]
Comment (m-nagarajan) [reply services/venice-server/src/test/java/com/linkedin/venice/stats/RocksDBStatsOtelTest.java:68]: Fixed. `setUp()` now creates a dedicated `AsyncGauge.AsyncGaugeExecutor` and wires it into the repository via `setTehutiMetricConfig(new MetricConfig(asyncGaugeExecutor))`, so `tearDown()`'s `metricsRepository.close()` only shuts down the per-test executor — not the JVM-wide singleton:

```java
private AsyncGauge.AsyncGaugeExecutor asyncGaugeExecutor;

@BeforeMethod
public void setUp() throws IOException {
  ...
  asyncGaugeExecutor = new AsyncGauge.AsyncGaugeExecutor.Builder().build();
  metricsRepository = new VeniceMetricsRepository(
      new VeniceMetricsConfig.Builder()
          ...
          .setTehutiMetricConfig(new MetricConfig(asyncGaugeExecutor))
          .build());
  ...
}

@AfterMethod
public void tearDown() throws IOException {
  if (metricsRepository != null) {
    metricsRepository.close();
  }
  if (asyncGaugeExecutor != null) {
    asyncGaugeExecutor.close();
  }
}
```

Same pattern as `StoreBufferServiceStatsOtelTest`. Later AsyncGauge-based tests in the same JVM are unaffected.

File: services/venice-server/src/test/java/com/linkedin/venice/stats/RocksDBStatsOtelTest.java [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+
+  @Test
+  public void testMetricsBehaviorBeforeStatSet() {
+    InMemoryMetricReader uninitReader = InMemoryMetricReader.create();
+    try (VeniceMetricsRepository uninitRepo = new VeniceMetricsRepository(
+        new VeniceMetricsConfig.Builder().setMetricPrefix(TEST_METRIC_PREFIX)
+            .setMetricEntities(SERVER_METRIC_ENTITIES)
+            .setEmitOtelMetrics(true)
+            .setOtelAdditionalMetricsReader(uninitReader)
+            .build())) {
```
[the remark this answers is not in the document]
Comment (m-nagarajan) [reply services/venice-server/src/test/java/com/linkedin/venice/stats/RocksDBStatsOtelTest.java]: Fixed. `testMetricsBehaviorBeforeStatSet` now declares its own local `AsyncGauge.AsyncGaugeExecutor` inside the try-with-resources and wires it via `setTehutiMetricConfig(new MetricConfig(localExecutor))`, so closing the block only shuts down the per-test executor:

```java
try (AsyncGauge.AsyncGaugeExecutor localExecutor = new AsyncGauge.AsyncGaugeExecutor.Builder().build();
    VeniceMetricsRepository uninitRepo = new VeniceMetricsRepository(
        new VeniceMetricsConfig.Builder().setMetricPrefix(TEST_METRIC_PREFIX)
            .setMetricEntities(SERVER_METRIC_ENTITIES)
            .setEmitOtelMetrics(true)
            .setOtelAdditionalMetricsReader(uninitReader)
            .setTehutiMetricConfig(new MetricConfig(localExecutor))
            .build())) {
  ...
}
```

The JVM-wide static `DEFAULT_ASYNC_GAUGE_EXECUTOR` stays alive for any AsyncGauge-based test that runs afterward. Same fix applied to the other two ad-hoc repository constructions in this file (`testNoNpeWhenOtelDisabled` and the new `testOtelEmitsForTotalStatsName`).

File: services/venice-server/src/test/java/com/linkedin/venice/stats/RocksDBStatsOtelTest.java [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+            "VeniceRocksDBBlockCacheComponent." + component + " must emit a data point for " + metricName
+                + " — add it to the corresponding ticker map in RocksDBStats.");
+      }
+    }
+  }
+
+  @Test
+  public void testNoNpeWhenOtelDisabled() {
+    try (VeniceMetricsRepository disabledRepo = new VeniceMetricsRepository(
+        new VeniceMetricsConfig.Builder().setMetricPrefix(TEST_METRIC_PREFIX).setEmitOtelMetrics(false).build())) {
```
[the remark this answers is not in the document]
Comment (m-nagarajan) [reply services/venice-server/src/test/java/com/linkedin/venice/stats/RocksDBStatsOtelTest.java]: Fixed. `testNoNpeWhenOtelDisabled` now declares its own local `AsyncGauge.AsyncGaugeExecutor` inside the try-with-resources and wires it via `setTehutiMetricConfig(new MetricConfig(localExecutor))`, so closing the block only affects the per-test executor:

```java
try (AsyncGauge.AsyncGaugeExecutor localExecutor = new AsyncGauge.AsyncGaugeExecutor.Builder().build();
    VeniceMetricsRepository disabledRepo = new VeniceMetricsRepository(
        new VeniceMetricsConfig.Builder().setMetricPrefix(TEST_METRIC_PREFIX)
            .setEmitOtelMetrics(false)
            .setTehutiMetricConfig(new MetricConfig(localExecutor))
            .build())) {
  ...
}
```

The JVM-wide static `DEFAULT_ASYNC_GAUGE_EXECUTOR` is left untouched.

File: services/venice-server/src/main/java/com/linkedin/venice/stats/RocksDBStats.java [hunk 8]
```diff
@@ -28,112 +45,273 @@
...
+    // --- Block Cache Hit Ratio: Tehuti-only (OTel derivable: hit{data} / (hit{data} + sum(miss))) ---
+    registerSensorIfAbsent(new AsyncGauge((ig, ig2) -> {
       if (rocksDBStat != null) {
-        return rocksDBStat.getTickerCount(BLOCK_CACHE_DATA_HIT)
-            / (double) (rocksDBStat.getTickerCount(BLOCK_CACHE_DATA_HIT)
-                + rocksDBStat.getTickerCount(BLOCK_CACHE_MISS));
+        long dataHit = rocksDBStat.getTickerCount(BLOCK_CACHE_DATA_HIT);
+        long miss = rocksDBStat.getTickerCount(BLOCK_CACHE_MISS);
+        long denominator = dataHit + miss;
+        return denominator == 0 ? Double.NaN : dataHit / (double) denominator;
```
[the remark this answers is not in the document]
Comment (m-nagarajan) [reply services/venice-server/src/main/java/com/linkedin/venice/stats/RocksDBStats.java:238]: Pre-existing formula carried over verbatim from main — only refactored to add the divide-by-zero guard

**Review by sushantmane — APPROVED** [review]
